### ISSUE TITLE
Inconsistent key type inference when spreading objects with const assertion into plain objects

### ISSUE THREAD

**Issue by Not-Jayden** [microsoft/TypeScript#59102]
### 🔎 Search Terms

const assertion, type inference, as const, object keys, type conversion, keyof typeof, satisfies, spread, readonly object

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about keyof, as const, infererence etc.

### ⏯ Playground Link

[Playground](https://www.typescriptlang.org/play/?ts=5.5.2#code/MYewdgzgLgBGCGBbAphAQgTwJIBMYF4YBvAKBnJgEYAuGAIgCkQALMGAERGToBoyKATLUbwwyGAGVEASyjNe-cgGZhAQQA2yAB4wmrCOAUUYAFmEBZeACdp8GAHFrwWwoC+MeBBihIsCPChpCAAzaVQYACVkUCscAB4wAFdEACNkKx4YaBswAHMAPgBuEhIoDAAHcVwvQgBrZAwQYJgyyqa4JFRMXGKAel6KAD0AfhKfaA6UdGwcAGEQcowCYkUYADoNhCnunD53f0CQsK8omPik1PTM7Ok8opLWqpwIecXl+sbmx-atrpnXjB9AbkEZAA)

### 💻 Code

```ts
const namesById = {
  1: "John Doe",
  2: "Jane Smith",
  3: "Alex Johnson",
  4: "Maria Garcia",
} as const satisfies Record<number, string>;

type Ids = keyof typeof namesById;
//   ^? type Ids = 1 | 2 | 3 | 4

const namesByIdCopy = {
  ...namesById,
} satisfies Record<number, string>;

type IdsCopy = keyof typeof namesByIdCopy;
//   ^? type IdsCopy = "1" | "2" | "3" | "4"

```

### 🙁 Actual behavior

`IdsCopy` has converted all of the keys to string literal union instead of a number literal union when using `keyof typeof`, even despite the fact that we've used `satisfies Record<number, string>` on the `namesByIdCopy` object.

### 🙂 Expected behavior

```ts
type IdsCopy = keyof typeof namesByIdCopy;
//   ^? type IdsCopy = "1" | "2" | "3" | "4"
```

should instead be typed as:
```ts
type IdsCopy = keyof typeof namesByIdCopy;
//   ^? type IdsCopy = 1 | 2 | 3 | 4
```

### Additional information about the issue

The type inference seems to work as expected (i.e. numeric literal union) if `as const` is used on both the original object and the copy object, or if `as const` isn't used on the original object. It's only when spreading the object with  an`as const` and inferring the keys using `keyof typeof` that they get inferred incorrectly.

**Comment by Not-Jayden** [comment]
[Workbench repro](https://www.typescriptlang.org/dev/bug-workbench/?ts=5.4.0-dev.20240210#code/MYewdgzgLgBGCGBbAphAQgTwJIBMYF4YBvAKBnJgEYAuGAIgCkQALMGAERGToBoyKATLUbwwyGAGVEASyjNe-cgGZhAQQA2yAB4wmrCOAUUYAFmEBZeACdp8GAHFrwWwoC+MeBBihIsCPChpCAAzaVQYACVkUCscAB4wAFdEACNkKx4YaBswAHMAPgBuEhIoDAAHcVwvQgBrZAwQYJgyyqa4JFRMXGKAel6KAD0AfhKfaA6UdGwcAGEQcowCYkUYADoNhCnunD53f0CQsK8omPik1PTM7Ok8opLWqpwIecXl+sbmx-atrpnXjB9AbkEZAA)

**Comment by Not-Jayden** [comment]
```ts repro
const namesById = {
    1: "John Doe",
    2: "Jane Smith",
    3: "Alex Johnson",
    4: "Maria Garcia",
} as const satisfies Record&lt;number, string>;

type Ids = keyof typeof namesById;
//   ^?

const namesByIdCopy = {
    ...namesById,
} satisfies Record&lt;number, string>;

type IdsCopy = keyof typeof namesByIdCopy;
//   ^?
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?noImplicitAny=false#code/MYewdgzgLgBGCGBbAphAQgTwJIBMYF4YBvAKBnJgEYAuGAIgCkQALMGAERGToBoyKATLUbwwyGAGVEASyjNe-cgGZhAQQA2yAB4wmrCOAUUYAFmEBZeACdp8GAHFrwWwoC+MeBBihIsCPChpCAAzaVQYACVkUCscAB4wAFdEACNkKx4YaBswAHMAPgBuEhIoDAAHcVwvQgBrZAwQYJgyyqa4JFRMXGKAel6KAD0AfhKfaA6UdGwcAGEQcowCYkUYADoNhCnunD53f0CQsK8omPik1PTM7Ok8opLWqpwIecXl+sbmx-atrpnXjB9AbkEZAA)

**Comment by Not-Jayden** [comment]
```ts repro
const namesById = {
    1: "John Doe",
    2: "Jane Smith",
    3: "Alex Johnson",
    4: "Maria Garcia",
} as const satisfies Record<number, string>;

type Ids = keyof typeof namesById;
//   ^?

const namesByIdCopy = {
    ...namesById,
} satisfies Record<number, string>;

type IdsCopy = keyof typeof namesByIdCopy;
//   ^?
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?noImplicitAny=false#code/MYewdgzgLgBGCGBbAphAQgTwJIBMYF4YBvAKBnJgEYAuGAIgCkQALMGAERGToBoyKATLUbwwyGAGVEASyjNe-cgGZhAQQA2yAB4wmrCOAUUYAFmEBZeACdp8GAHFrwWwoC+MeBBihIsCPChpCAAzaVQYACVkUCscAB4wAFdEACNkKx4YaBswAHMAPgBuEhIoDAAHcVwvQgBrZAwQYJgyyqa4JFRMXGKAel6KAD0AfhKfaA6UdGwcAGEQcowCYkUYADoNhCnunD53f0CQsK8omPik1PTM7Ok8opLWqpwIecXl+sbmx-atrpnXjB9AbkEZAA)

**Comment by MichalMarsalek** [comment]
I'm slightly confused about what the `satisfies` has to do with anything. Doesn't

```ts
const namesById = {
    1: "John Doe",
    2: "Jane Smith",
    3: "Alex Johnson",
    4: "Maria Garcia",
} as const;
type Ids = keyof typeof namesById;
const namesByIdCopy = {...namesById};
type IdsCopy = keyof typeof namesByIdCopy;
```
demonstrate the issue sufficiently?

**Comment by Not-Jayden** [comment]
> I'm slightly confused about what the `satisfies` has to do with anything. Doesn't
> 
> ```ts
> const namesById = {
>     1: "John Doe",
>     2: "Jane Smith",
>     3: "Alex Johnson",
>     4: "Maria Garcia",
> } as const;
> type Ids = keyof typeof namesById;
> const namesByIdCopy = {...namesById};
> type IdsCopy = keyof typeof namesByIdCopy;
> ```
> 
> demonstrate the issue sufficiently?

@MichalMarsalek Yeah the satisfies isn't necessarily important/related to the underlying issue, it just further illustrates the point that `namesByIdCopy` does supposedly have a number key, but doesn't when you use `keyof typeof`
